src/Furigana.cpp

```diff
@@ -0,0 +1,169 @@
+
+#include <cstring>
+#include <algorithm>
+#include <string.h>
+
+#include "Furigana.h"
+#include "Utf8.h"
+
+std::unordered_map<uint32_t, uint32_t> Furigana::KataToHiraTable;
+
+const char Furigana::katakanas[][4] = {
+    "ァ","ア","ィ","イ","ゥ","ウ","ェ","エ","ォ","オ",
+    "カ","ガ","キ","ギ","ク","グ","ケ","ゲ","コ","ゴ",
+    "サ","ザ","シ","ジ","ス","ズ","セ","ゼ","ソ","ゾ",
+    "タ","ダ","チ","ヂ","ッ","ツ","ヅ","テ","デ","ト",
+    "ド","ナ","ニ","ヌ","ネ","ノ","ハ","バ","パ","ヒ",
+    "ビ","ピ","フ","ブ","プ","ヘ","ベ","ペ","ホ","ボ",
+    "ポ","マ","ミ","ム","メ","モ","ャ","ヤ","ュ","ユ",
+    "ョ","ヨ","ラ","リ","ル","レ","ロ","ヮ","ワ","ヲ",
+    "ン","ヴ","ヵ","ヶ"
+};
+
+const char Furigana::hiraganas[][4] = {
+    "ぁ","あ","ぃ","い","ぅ","う","ぇ","え","ぉ","お",
+    "か","が","き","ぎ","く","ぐ","け","げ","こ","ご",
+    "さ","ざ","し","じ","す","ず","せ","ぜ","そ","ぞ",
+    "た","だ","ち","ぢ","っ","つ","づ","て","で","と",
+    "ど","な","に","ぬ","ね","の","は","ば","ぱ","ひ",
+    "び","ぴ","ふ","ぶ","ぷ","へ","べ","ぺ","ほ","ぼ",
+    "ぽ","ま","み","む","め","も","ゃ","や","ゅ","ゆ",
+    "ょ","よ","ら","り","る","れ","ろ","ゎ","わ","を",
+    "ん","ゔ","ゕ","ゖ"
+};
+
+Furigana::Furigana() {
+    /* Initialize the katakana to hiragana hash table. */
+    static size_t numKanas = sizeof(Furigana::katakanas)/sizeof(Furigana::katakanas[0]);
+    for (int i = 0; i < numKanas; i++) {
+        uint32_t katakana = *(uint32_t*)Furigana::katakanas[i];
+        uint32_t hiragana = *(uint32_t*)Furigana::hiraganas[i];
+        Furigana::KataToHiraTable.insert(std::make_pair(katakana, hiragana));
+    }
+}
+
+/**
+ * Return the given string with katakanas converted
+ * to their hiragana equivalents.
+ */
+std::string Furigana::katakana_to_hiragana(std::string katakana)
+{
+    const char *str = katakana.c_str();
+    char character[5] = { '\0' }; /* 4 UTF-8 bytes + '\0' = 5 */
+    std::string hiraganas;
+
+    while (utf8_getc(&str, character, sizeof(character)-1)) {
+        auto res = Furigana::KataToHiraTable.find(*(uint32_t*)character);
+        if (res != Furigana::KataToHiraTable.end()) {
+            std::memcpy(character, &res->second, sizeof(character)-1);
+        }
+
+        hiraganas.append(character, std::strlen(character));
+    }
+
+    return hiraganas;
+}
+
+static inline void remove_spaces(std::string &str) {
+   str.erase(std::remove_if(str.begin(), str.end(), (int(*)(int))std::isspace), str.end());
```

Comment:
for long lines can you indent  long lines / or lines with lot of parameters like this  (the original intent is 4 spaces by indenta

```
   str.erase(
   std::remove_if(
             str.begin(),
                 str.end(), (int(*)(int))std::isspace)
```